src/main/java/com/nerodesk/om/aws/AwsDoc.java

```diff
@@ -27,81 +27,88 @@
 
 /**
- * Amazon S3 storage.
+ * AWS-based version of Doc.
  *
- * @author Alexey Saenko (alexey.saenko@gmail.com)
+ * @author Yegor Bugayenko (yegor@teamed.io)
  * @version $Id$
  * @since 0.2
  */
 @ToString
 @EqualsAndHashCode
-public final class AmazonStorage implements Storage {
+public final class AwsDoc implements Doc {
 
     /**
-     * Region.
+     * Bucket.
      */
-    private final transient Region region;
+    private final transient Bucket bucket;
 
     /**
-     * Bucket name.
+     * URN of user.
      */
-    private final transient String bucket;
+    private final transient String user;
 
     /**
-     * Ctor.
-     * @param key Amazon key
-     * @param secret Amazon secret
-     * @param bckt The name of the bucket to get
+     * Doc name.
      */
-    public AmazonStorage(@NotNull final String key,
-        @NotNull final String secret, @NotNull final String bckt) {
-        this(new Region.Simple(key, secret), bckt);
-    }
+    private final transient String label;
 
     /**
      * Ctor.
-     * @param rgn Amazon region abstraction
-     * @param bckt The name of the bucket to get
+     * @param bkt Bucket
+     * @param urn User URN
+     * @param doc Name of document
      */
-    public AmazonStorage(
-        @NotNull final Region rgn,
-        @NotNull final String bckt) {
-        super();
-        this.region = rgn;
-        this.bucket = bckt;
+    public AwsDoc(final Bucket bkt, final String urn, final String doc) {
+        this.bucket = bkt;
+        this.user = urn;
+        this.label = doc;
+    }
+
+    @Override
+    public boolean exists() throws IOException {
+        return this.ocket().exists();
     }
 
     @Override
-    public InputStream get(@NotNull final String path) throws IOException {
-        final PipedOutputStream pos = new PipedOutputStream();
-        this.region.bucket(this.bucket).ocket(path).read(pos);
-        return new PipedInputStream(pos);
+    public void delete() throws IOException {
+        this.bucket.remove(this.ocket().key());
     }
 
     @Override
-    public void put(
-        @NotNull final String path,
-        @NotNull final InputStream input) throws IOException {
-        final Ocket ocket = this.region.bucket(this.bucket).ocket(path);
-        ocket.write(input, ocket.meta());
+    public void read(@NotNull final OutputStream output) throws IOException {
+        this.ocket().read(output);
+        Logger.info(this, "%s read", this.label);
     }
 
     @Override
-    public void delete(@NotNull final String path) throws IOException {
-        this.region.bucket(this.bucket).remove(path);
+    public void write(final InputStream input) throws IOException {
+        this.ocket().write(input, new ObjectMetadata());
+        Logger.info(this, "%s written", this.label);
+    }
+
+    /**
+     * Get ocket.
+     * @return Ocket
+     */
+    private Ocket ocket() {
+        return this.bucket.ocket(
+            String.format("/%s%s", this.user, this.label)
```

Comment:
Does it guarantee unique key?
For example below 2 different user/files that produces the same key.
 
```
user="urn:test:1";
label="23.txt"
```

and
```
user="urn:test:12";
label="3.txt"
```